### PR TITLE
Sync Pagination and fix bugs

### LINKED ISSUE #45: [Syncing] Pagination
## Points to note
- Component in action - `Pagination.js`
- Github Urls to this component:
  * [neeto-ui/Pagination.js](https://github.com/bigbinary/neeto-ui/tree/67ee4a5188c5feccbeca2767107651fb24932764/lib/components/Pagination.js)
  * [NitroUI/Pagination.js](https://github.com/bigbinary/neeto-help-web/tree/0b11a7b0f98ac7c533df1e9ec37cf900b4cbc901/app/javascript/Admin/Components/NitroUI/Pagination.js)
- :green_circle: Green color signifies how that particular line is written in `neeto-ui` package
- :red_circle: Red color signifies how that particular line is written in `NitroUI` internals

## Sync Required or Not
- Are there significant changes for `Pagination.js` that needs to be synced to `neeto-ui` package from `NitroUI`? - **No**
- I'm not sure whether the `pseudoClasses` and `disableNavigateBack` are needed. Rather couldn't the `lastPage` and `pageNo===1` in `neetoui` package bring in the same functionality?

## Diff
```diff
--- NitroUI/Pagination.js	2020-11-28 16:54:24.002184800 +0530
+++ neetoui-official/Pagination.js	2020-11-28 16:49:18.749268700 +0530
@@ -3,7 +3,7 @@
 import classnames from "classnames";
 
 const LeftSVG = () => (
-  <svg className="w-4.5 h-4.5" viewBox="0 0 20 20" fill="currentColor">
+  <svg className="w-5 h-5" viewBox="0 0 20 20" fill="currentColor">
     <path
       fillRule="evenodd"
       d="M12.707 5.293a1 1 0 010 1.414L9.414 10l3.293 3.293a1 1 0 01-1.414 1.414l-4-4a1 1 0 010-1.414l4-4a1 1 0 011.414 0z"
@@ -13,7 +13,7 @@
 );
 
 const RightSVG = () => (
-  <svg className="w-4.5 h-4.5" viewBox="0 0 20 20" fill="currentColor">
+  <svg className="w-5 h-5" viewBox="0 0 20 20" fill="currentColor">
     <path
       fillRule="evenodd"
       d="M7.293 14.707a1 1 0 010-1.414L10.586 10 7.293 6.707a1 1 0 011.414-1.414l4 4a1 1 0 010 1.414l-4 4a1 1 0 01-1.414 0z"
@@ -30,7 +30,7 @@
     end = count < pageNo * pageSize ? count : pageNo * pageSize;
   }
   let lastPage = false;
-  if (count <= pageNo * pageSize && count > (pageNo - 1) * pageSize) {
+  if (count < pageNo * pageSize && count > (pageNo - 1) * pageSize) {
     lastPage = true;
   }
 
@@ -38,20 +38,6 @@
     const lastPageIndex = parseInt(count / pageSize) + 1;
     navigate(lastPageIndex);
   };
-
-  const baseClasses = [
-    "relative inline-flex items-center px-2 py-2 text-sm font-medium leading-5 text-gray-400",
-    "transition duration-150 ease-in-out bg-white border border-gray-300 focus:outline-none"
-  ];
-
-  const psuedoClasses = [
-    "hover:bg-gray-50 hover:text-gray-700 active:bg-gray-100 active:text-gray-700"
-  ];
-  const disabledClasses = ["cursor-not-allowed bg-gray-100"];
-
-  const disableNavigateBack = count === 0 || pageNo === 1;
-  const disableNavigateForward = count === 0 || lastPage;
-
   return (
     <div
       className={classnames([
@@ -61,7 +47,7 @@
     >
       <Dropdown
         customTarget={() => (
-          <p className="text-gray-600 cursor-pointer">
+          <p className="text-gray-600">
             <span className="font-semibold text-gray-800">
               {start} - {end}
             </span>{" "}
@@ -76,25 +62,28 @@
       </Dropdown>
       <div className="relative z-0 inline-flex ml-3 shadow-sm">
         <button
-          className={classnames(baseClasses, "rounded-l-md border-r-0", {
-            [psuedoClasses]: !disableNavigateBack,
-            [disabledClasses]: disableNavigateBack
-          })}
+          className={classnames(
+            [
+              "relative inline-flex items-center px-2 py-2 text-sm font-medium leading-5 text-gray-400 transition duration-150 ease-in-out bg-white border border-gray-300 rounded-l-md hover:text-gray-700 focus:z-10 focus:outline-none hover:bg-gray-100 active:bg-gray-100 active:text-gray-700"
+            ],
+            { "cursor-not-allowed": pageNo === 1 }
+          )}
           aria-label="Previous"
           onClick={() => navigate(pageNo - 1)}
-          disabled={disableNavigateBack}
-          disabled={disableNavigateBack}
+          disabled={pageNo === 1}
         >
           <LeftSVG />
         </button>
         <button
-          className={classnames(baseClasses, "rounded-r-md", {
-            [psuedoClasses]: !disableNavigateForward,
-            [disabledClasses]: disableNavigateForward
-          })}
+          className={classnames(
+            [
+              "relative inline-flex items-center px-2 py-2 -ml-px text-sm font-medium leading-5 text-gray-400 transition duration-150 ease-in-out bg-white border border-gray-300 rounded-r-md hover:text-gray-700 focus:z-10 focus:outline-none hover:bg-gray-100 active:bg-gray-100 active:text-gray-700"
+            ],
+            { "cursor-not-allowed": lastPage }
+          )}
           aria-label="Next"
           onClick={() => navigate(pageNo + 1)}
-          disabled={disableNavigateForward}
+          disabled={lastPage}
         >
           <RightSVG />
         </button>
```

### PR DESCRIPTION
**Pull request by yedhink**
Fixes #45 

I have fixed some bugs in the component.

### Please check out this demo before merging
- https://youtu.be/kd5ox6Bdkk4
- Also made `pseudoClasses` and `disabledClasses` just string rather than array(in neeto-help it's array), since we are using it as key.
- Also added a check if `count !==0` for making jump to last and first clickable.

@vinay0x _a
Would you be able to review this PR and see if everything is as intended? Thanks

cc: @goutham-subramanyam

[Neeto-Help component reference](https://github.com/bigbinary/neeto-help-web/blob/4075f35979924ca64d6727c97c82a8f61d9ba739/app/javascript/Admin/Components/NitroUI/Pagination.js)

### CODE REVIEW

File: lib/components/Pagination.js [hunk 1]
```diff
@@ -2,42 +2,30 @@ import React from "react";
...
   }
-  let lastPage = false;
-  if (count < pageNo * pageSize && count > (pageNo - 1) * pageSize) {
-    lastPage = true;
-  }
+  const isLastPage = pageNo * pageSize === count;
 
   const jumpToLast = () => {
-    const lastPageIndex = parseInt(count / pageSize) + 1;
+    const lastPageIndex = Math.floor(count / pageSize);
```
Comment (goutham-subramanyam) [inline lib/components/Pagination.js:14]: Why were these logics updated?

**Review by goutham-subramanyam — COMMENTED** [review]
@yedhink _a

File: lib/components/Pagination.js (hunk 1 above)
Comment (yedhink) [reply lib/components/Pagination.js:14]: @goutham-subramanyam 

`parseInt` is for parsing a string into integer. 

I didn't understand the point of having a `+ 1`. Isn't it a bug?

Eg: Let's say total count = 10. And pageSize = 5. Then the last page index = 2. Not `2+1=3`.

Do let me know if there is some other use case for `+ 1`.

**Comment by yedhink** [comment]
There are some more minor bugs with this. Don't merge this yet.

**Comment by yedhink** [comment]
@goutham-subramanyam _a

- Have cleaned up the code
- Fixed the bug that occurred when we gave values like pageSize=5 and count=17.

Please review and merge if everything looks good.

**Comment by goutham-subramanyam** [comment]
@edwinbbu _a

File: lib/components/Pagination.js (hunk 1 above)
Comment (edwinbbu) [reply lib/components/Pagination.js]: @yedhink _a Lets say we are in page 1. The total count = 9. pageSize = 5. We are calling the jumpToLast function.
(count/pageSize) provides 9/5 = 1 which is still in page 1. Hence we are providing +1.

File: src/index.js [hunk 2]
```diff
@@ -68,6 +70,13 @@ const App = () => {
         minDate={moment("2019-01-01").toDate()}
         maxDate={moment().toDate()}
       />
+
+      <Pagination
+        count={17}
+        pageNo={pageNo}
+        pageSize={PAGE_SIZE}
+        navigate={index => setPageNo(index)}
```
Comment (edwinbbu) [inline src/index.js:78]: Also test `jumpToFirst` and `jumpToLast` conditions.

**Comment by yedhink** [comment]
@edwinbbu _a

Please check this demo and see if it answers your questions: https://youtu.be/WzI927fuERY

If everything looks good, I think we can merge this.

File: lib/components/Pagination.js [hunk 3]
```diff
@@ -2,42 +2,24 @@ import React from "react";
...
-      clipRule="evenodd"
-    />
-  </svg>
-);
+const Pagination = ({ count = 0, pageNo, pageSize, navigate, className = "", emptyPageMsg = "No records" }) => {
+  const maxPageNo = Math.ceil(count / pageSize);
+  const isPageEmpty = count === 0;
+  const startIndexOfCurrentPage = isPageEmpty ? 0 : ((pageNo - 1) * pageSize) + 1;
+  const endIndexOfCurrentPage = Math.min(count, pageNo * pageSize);
+  const isLastPage = endIndexOfCurrentPage >= count;
```
Comment (vinay0x) [inline lib/components/Pagination.js:10]: Why use count when you can do `pageNo == maxPageNo`?

File: lib/components/Pagination.js [hunk 4]
```diff
@@ -2,42 +2,24 @@ import React from "react";
...
-      d="M12.707 5.293a1 1 0 010 1.414L9.414 10l3.293 3.293a1 1 0 01-1.414 1.414l-4-4a1 1 0 010-1.414l4-4a1 1 0 011.414 0z"
-      clipRule="evenodd"
-    />
-  </svg>
-);
+const Pagination = ({ count = 0, pageNo, pageSize, navigate, className = "", emptyPageMsg = "No records" }) => {
+  const maxPageNo = Math.ceil(count / pageSize);
+  const isPageEmpty = count === 0;
+  const startIndexOfCurrentPage = isPageEmpty ? 0 : ((pageNo - 1) * pageSize) + 1;
+  const endIndexOfCurrentPage = Math.min(count, pageNo * pageSize);
```
Comment (vinay0x) [inline lib/components/Pagination.js:9]: Is this logic really needed? In what case will count be greater than `pageNo * pageSize`?

File: lib/components/Pagination.js [hunk 5]
```diff
@@ -2,42 +2,24 @@ import React from "react";
...
-    end = 0;
-  if (count) {
-    start = pageNo > 1 ? (pageNo - 1) * pageSize + 1 : 1;
-    end = count < pageNo * pageSize ? count : pageNo * pageSize;
-  }
-  let lastPage = false;
-  if (count < pageNo * pageSize && count > (pageNo - 1) * pageSize) {
-    lastPage = true;
-  }
+  const psuedoClasses = "hover:bg-gray-50 hover:text-gray-700 active:bg-gray-100 active:text-gray-700";
```
Comment (vinay0x) [inline lib/components/Pagination.js:17]: If the other one is `disabledClasses`, name this `activeClasses`.

**Review by vinay0x — CHANGES_REQUESTED** [review]
@yedhink _a

**Comment by yedhink** [comment]
@vinay0x _a

Have made the changes. Please review and merge if everything looks good. Thanks.

**Pull request merged**